File: .cg-cache/process.js
const fs = require('fs');
const markets = JSON.parse(fs.readFileSync('.cg-cache/markets.json', 'utf8'));
const trending = JSON.parse(fs.readFileSync('.cg-cache/trending.json', 'utf8'));

const STABLE_IDS = new Set(['tether','usd-coin','dai','first-digital-usd','usde','tusd','usdd','pyusd','fdusd','paxg','usds','ethena-usde','frax','usdt0','susds','ondo-us-dollar-yield','blackrock-usd-institutional-digital-liquidity-fund','sky-dollar','resolv-usr','falcon-usd','m0-foundation','tether-eurt','usdx-money-usdx','tether-gold','staked-frax','crypto-com-staked-eth','ethena-staked-usde','usdq','usd0','dola-usd','first-digital-labs-usd','usdtb']);
const WRAPPED_IDS = new Set(['wrapped-bitcoin','wrapped-steth','staked-ether','weth','wbeth','wrapped-eeth','rocket-pool-eth','coinbase-wrapped-btc','lombard-staked-btc','jito-staked-sol','binance-staked-sol','marinade-staked-sol','liquid-staked-ether','solv-protocol-solvbtc','renzo-restaked-eth','kelp-dao-restaked-eth','mantle-staked-ether','mantle-restaked-eth','tbtc','m-by-m0','ondo-us-dollar-yield','staked-frax-ether','stader-ethx','frax-ether','rocket-pool-eth','sky-dollar']);

function isStable(c) {
  if (STABLE_IDS.has(c.id)) return true;
  const sym = (c.symbol||'').toUpperCase();
  const name = (c.name||'').toLowerCase();
  if (sym.startsWith('USD')||sym.startsWith('EUR')||sym.startsWith('GBP')) return true;
  if (name.includes('stablecoin')||name.includes('stable coin')) return true;
  const p = c.current_price||0;
  if (p>=0.97 && p<=1.03 && Math.abs(c.price_change_percentage_24h||0)<0.5 && (c.market_cap_rank||9999)<=200) return true;
  return false;
}

const filtered = markets.filter(c => !isStable(c) && !WRAPPED_IDS.has(c.id) && (c.total_volume||0)>=1_000_000 && c.price_change_percentage_24h != null);

const trendingIds = new Set();
const trendingSyms = new Set();
for (const t of (trending.coins||[])) {
  const item = t.item||{};
  trendingIds.add(item.id);
  trendingSyms.add((item.symbol||'').toLowerCase());
}

function tagsFor(c, inTrending, listName) {
  const tags = [];
  const p24 = c.price_change_percentage_24h||0;
  const p7d = c.price_change_percentage_7d_in_currency||0;
  const rank = c.market_cap_rank||9999;
  const mcap = c.market_cap||0;
  const vol = c.total_volume||0;
  const volMcap = mcap>0 ? vol/mcap : 0;
  if (inTrending && listName==='win') tags.push('TRENDING+UP');
  if (inTrending && listName==='lose') tags.push('TRENDING+DOWN');
  if (p24>15 && p7d>25) tags.push('BREAKOUT');
  if (p24>20 && p7d<0) tags.push('FADE');
  if (p24<-10 && volMcap>0.25) tags.push('CAPITULATION');
  if (rank>150 && p24>30) tags.push('PUMP-RISK');
  if (mcap < 50_000_000) tags.push('MICROCAP');
  if (rank<=20) tags.push('MAJOR');
  return tags.slice(0,2);
}

const winners = [...filtered].sort((a,b)=> (b.price_change_percentage_24h||0)-(a.price_change_percentage_24h||0)).slice(0,10);
const losers = [...filtered].sort((a,b)=> (a.price_change_percentage_24h||0)-(b.price_change_percentage_24h||0)).slice(0,10);

const top100 = filtered.slice(0,100);
const green = top100.filter(c=>(c.price_change_percentage_24h||0)>0).length;
const top50ch = filtered.slice(0,50).map(c=>c.price_change_percentage_24h||0).sort((a,b)=>a-b);
const median = top50ch.length ? top50ch[Math.floor(top50ch.length/2)] : 0;

function fmtPrice(p){
  if (p==null) return '?';
  if (p>=1000) return '$'+p.toLocaleString(undefined,{maximumFractionDigits:0});
  if (p>=1) return '$'+p.toFixed(2);
  if (p>=0.01) return '$'+p.toFixed(4);
  if (p>=0.0001) return '$'+p.toFixed(6);
  return '$'+p.toFixed(8);
}
function fmtAbbr(v){
  if (v==null) return '?';
  if (v>=1e9) return '$'+(v/1e9).toFixed(1)+'B';
  if (v>=1e6) return '$'+(v/1e6).toFixed(0)+'M';
  if (v>=1e3) return '$'+(v/1e3).toFixed(0)+'K';
  return '$'+v.toFixed(0);
}

function render(c, listName){
  const inTrending = trendingIds.has(c.id) || trendingSyms.has((c.symbol||'').toLowerCase());
  const tg = tagsFor(c, inTrending, listName);
  return {
    symbol: (c.symbol||'').toUpperCase(),
    name: c.name,
    id: c.id,
    price: fmtPrice(c.current_price),
    p24: c.price_change_percentage_24h||0,
    p7d: c.price_change_percentage_7d_in_currency||0,
    p1h: c.price_change_percentage_1h_in_currency||0,
    vol: fmtAbbr(c.total_volume),
    rank: c.market_cap_rank,
    tags: tg,
  };
}

const winnersR = winners.map(c=>render(c,'win'));
const losersR = losers.map(c=>render(c,'lose'));

console.log(`PULSE: ${green}/100 top-100 green, median_top50=${median.toFixed(2)}%`);
console.log('\n=== WINNERS ===');
for (const w of winnersR) {
  const tg = w.tags.length ? ` [${w.tags.join('/')}]` : '';
  console.log(`  ${w.symbol} (${w.name}) — ${w.price}  ${w.p24>=0?'+':''}${w.p24.toFixed(1)}% / 7d ${w.p7d>=0?'+':''}${w.p7d.toFixed(1)}% / 1h ${w.p1h>=0?'+':''}${w.p1h.toFixed(1)}%  •  ${w.vol} / #${w.rank}${tg}`);
}
console.log('\n=== LOSERS ===');
for (const l of losersR) {
  const tg = l.tags.length ? ` [${l.tags.join('/')}]` : '';
  console.log(`  ${l.symbol} (${l.name}) — ${l.price}  ${l.p24>=0?'+':''}${l.p24.toFixed(1)}% / 7d ${l.p7d>=0?'+':''}${l.p7d.toFixed(1)}% / 1h ${l.p1h>=0?'+':''}${l.p1h.toFixed(1)}%  •  ${l.vol} / #${l.rank}${tg}`);
}

console.log('\n=== TRENDING ===');
const trendingItems = (trending.coins||[]).slice(0,7).map(t=>{
  const item = t.item||{};
  const data = item.data||{};
  return {
    symbol: (item.symbol||'').toUpperCase(),
    name: item.name,
    id: item.id,
    rank: item.market_cap_rank,
    price: fmtPrice(data.price),
    change_24h: (data.price_change_percentage_24h||{}).usd,
  };
});
for (const t of trendingItems) {
  const ch = t.change_24h;
  const chs = ch!=null ? `${ch>=0?'+':''}${ch.toFixed(1)}%` : '?';
  const inWin = winnersR.some(w=>w.symbol===t.symbol);
  const inLose = losersR.some(l=>l.symbol===t.symbol);
  const ex = [];
  if (inWin) ex.push('TRENDING+UP');
  if (inLose) ex.push('TRENDING+DOWN');
  const tg = ex.length ? ` [${ex.join('/')}]` : '';
  console.log(`  ${t.symbol} (${t.name}) — #${t.rank}, ${t.price}, 24h ${chs}${tg}`);
}

fs.writeFileSync('.cg-cache/results.json', JSON.stringify({
  pulse:{green100:green, median_top50:median},
  winners:winnersR,
  losers:losersR,
  trending:trendingItems,
}, null, 2));
console.log('\nSaved .cg-cache/results.json');

File: .cg-cache/process.py
#!/usr/bin/env python3
import json

with open('.cg-cache/markets.json') as f:
    markets = json.load(f)
with open('.cg-cache/trending.json') as f:
    trending = json.load(f)

STABLE_IDS = {'tether','usd-coin','dai','first-digital-usd','usde','tusd','usdd','pyusd','fdusd','paxg','usds','ethena-usde','frax','usdt0','susds','ondo-us-dollar-yield','blackrock-usd-institutional-digital-liquidity-fund','sky-dollar','resolv-usr','falcon-usd','m0-foundation','tether-eurt','usdx-money-usdx','tether-gold','staked-frax'}
WRAPPED_IDS = {'wrapped-bitcoin','wrapped-steth','staked-ether','weth','wbeth','wrapped-eeth','rocket-pool-eth','coinbase-wrapped-btc','lombard-staked-btc','jito-staked-sol','binance-staked-sol','marinade-staked-sol','liquid-staked-ether','solv-protocol-solvbtc','renzo-restaked-eth','kelp-dao-restaked-eth','mantle-staked-ether','mantle-restaked-eth','tbtc','m-by-m0','susds','sky-dollar','ondo-us-dollar-yield','staked-frax-ether','rocket-pool','stader-ethx','frax-ether'}

def is_stable(c):
    if c['id'] in STABLE_IDS: return True
    sym = (c.get('symbol') or '').upper()
    name = (c.get('name') or '').lower()
    if sym.startswith('USD') or sym.startswith('EUR') or sym.startswith('GBP'): return True
    if 'stablecoin' in name or 'stable coin' in name: return True
    # peg-near-1 detector
    p = c.get('current_price') or 0
    if 0.97 <= p <= 1.03 and abs(c.get('price_change_percentage_24h') or 0) < 0.5 and c.get('market_cap_rank') and c['market_cap_rank'] <= 200:
        return True
    return False

filtered = []
for c in markets:
    if is_stable(c): continue
    if c['id'] in WRAPPED_IDS: continue
    vol = c.get('total_volume') or 0
    if vol < 1_000_000: continue
    if c.get('price_change_percentage_24h') is None: continue
    filtered.append(c)

# Tags
def tags_for(c, in_trending, list_name):
    tags = []
    p24 = c.get('price_change_percentage_24h') or 0
    p7d = c.get('price_change_percentage_7d_in_currency') or 0
    rank = c.get('market_cap_rank') or 9999
    mcap = c.get('market_cap') or 0
    vol = c.get('total_volume') or 0
    vol_mcap = vol/mcap if mcap > 0 else 0

    if in_trending and list_name == 'win':
        tags.append('TRENDING+UP')
    if in_trending and list_name == 'lose':
        tags.append('TRENDING+DOWN')
    if p24 > 15 and p7d > 25:
        tags.append('BREAKOUT')
    if p24 > 20 and p7d < 0:
        tags.append('FADE')
    if p24 < -10 and vol_mcap > 0.25:
        tags.append('CAPITULATION')
    if rank > 150 and p24 > 30:
        tags.append('PUMP-RISK')
    if mcap < 50_000_000:
        tags.append('MICROCAP')
    if rank <= 20:
        tags.append('MAJOR')
    return tags[:2]

trending_ids = set()
trending_syms = set()
for t in trending.get('coins', []):
    item = t.get('item', {})
    trending_ids.add(item.get('id'))
    trending_syms.add((item.get('symbol') or '').lower())

# Sort
winners = sorted(filtered, key=lambda c: c.get('price_change_percentage_24h') or 0, reverse=True)[:10]
losers = sorted(filtered, key=lambda c: c.get('price_change_percentage_24h') or 0)[:10]

# Trending
trending_items = []
for t in trending.get('coins', [])[:7]:
    item = t.get('item', {})
    data = item.get('data', {})
    trending_items.append({
        'name': item.get('name'),
        'symbol': (item.get('symbol') or '').upper(),
        'id': item.get('id'),
        'rank': item.get('market_cap_rank'),
        'price': data.get('price'),
        'change_24h': (data.get('price_change_percentage_24h') or {}).get('usd'),
    })

# Market pulse
top100 = filtered[:100]
green = sum(1 for c in top100 if (c.get('price_change_percentage_24h') or 0) > 0)
top50_changes = sorted([c.get('price_change_percentage_24h') or 0 for c in filtered[:50]])
median_top50 = top50_changes[len(top50_changes)//2] if top50_changes else 0
print(f"PULSE: {green}/100 top-100 green, median_top50={median_top50:.2f}")

def fmt_price(p):
    if p is None: return "?"
    if p >= 1000: return f"${p:,.0f}"
    if p >= 1: return f"${p:.2f}"
    if p >= 0.01: return f"${p:.4f}"
    if p >= 0.0001: return f"${p:.6f}"
    return f"${p:.8f}"

def fmt_abbr(v):
    if v is None: return "?"
    if v >= 1e9: return f"${v/1e9:.1f}B"
    if v >= 1e6: return f"${v/1e6:.0f}M"
    if v >= 1e3: return f"${v/1e3:.0f}K"
    return f"${v:.0f}"

def render(c, list_name):
    in_trending = c['id'] in trending_ids or c['symbol'].lower() in trending_syms
    tg = tags_for(c, in_trending, list_name)
    p24 = c.get('price_change_percentage_24h') or 0
    p7d = c.get('price_change_percentage_7d_in_currency') or 0
    p1h = c.get('price_change_percentage_1h_in_currency') or 0
    return {
        'symbol': c['symbol'].upper(),
        'name': c['name'],
        'price': fmt_price(c.get('current_price')),
        'p24': p24,
        'p7d': p7d,
        'p1h': p1h,
        'vol': fmt_abbr(c.get('total_volume')),
        'rank': c.get('market_cap_rank'),
        'tags': tg,
    }

print("\n=== WINNERS ===")
winners_r = [render(c, 'win') for c in winners]
for w in winners_r:
    tg = f" [{'/'.join(w['tags'])}]" if w['tags'] else ""
    print(f"  {w['symbol']} ({w['name']}) — {w['price']}  +{w['p24']:.1f}% / 7d {w['p7d']:+.1f}% / 1h {w['p1h']:+.1f}%  •  {w['vol']} / #{w['rank']}{tg}")

print("\n=== LOSERS ===")
losers_r = [render(c, 'lose') for c in losers]
for l in losers_r:
    tg = f" [{'/'.join(l['tags'])}]" if l['tags'] else ""
    print(f"  {l['symbol']} ({l['name']}) — {l['price']}  {l['p24']:.1f}% / 7d {l['p7d']:+.1f}% / 1h {l['p1h']:+.1f}%  •  {l['vol']} / #{l['rank']}{tg}")

print("\n=== TRENDING ===")
for t in trending_items:
    sym = t['symbol']
    name = t['name']
    rank = t['rank']
    price = fmt_price(t['price'])
    ch = t['change_24h']
    ch_str = f"{ch:+.1f}%" if ch is not None else "?"
    # tags for trending: check if it's also in winners/losers
    in_winners = any(w['symbol'] == sym for w in winners_r)
    in_losers = any(l['symbol'] == sym for l in losers_r)
    extras = []
    if in_winners: extras.append('TRENDING+UP')
    if in_losers: extras.append('TRENDING+DOWN')
    tg = f" [{'/'.join(extras)}]" if extras else ""
    print(f"  {sym} ({name}) — #{rank}, {price}, 24h {ch_str}{tg}")

# Save results for shell use
out = {
    'pulse': {'green100': green, 'median_top50': median_top50},
    'winners': winners_r,
    'losers': losers_r,
    'trending': [{'symbol': t['symbol'], 'name': t['name'], 'rank': t['rank'], 'price': fmt_price(t['price']), 'change_24h': t['change_24h']} for t in trending_items],
}
with open('.cg-cache/results.json', 'w') as f:
    json.dump(out, f, indent=2, default=str)
print("\nSaved .cg-cache/results.json")
